manticore/core/cpu/abstractcpu.py

```diff
@@ -549,89 +641,99 @@ def pop_int(self):
-                        break
-                assert isinstance(c, str)
-                text += c
-        except MemoryException:
-            pass
+        # Read Instruction from memory
+        for address in xrange(pc, pc+self.max_instr_width):
+            #This reads a byte from memory ignoring permissions
+            #and concretize it if symbolic
+            if not self.memory.access_ok(address, 'x'):
+                break
+
+            c = self.memory[address]
+
+            if issymbolic(c):
+                assert isinstance(c, BitVec) and  c.size == 8
+                if isinstance(c, Constant):
+                    c = chr(c.value)
+                else:
+                    logger.error('Concretize executable memory %r %r', c, text )
+                    raise ConcretizeMemory(self.memory, address = pc,
+                                            size = 8 * self.max_instr_width, 
+                                            policy = 'INSTRUCTION' )
+            text += c
+
+
+        #Pad potentially incomplete intruction with zeroes
 
         code = text.ljust(self.max_instr_width, '\x00')
-        instruction = next(self._md.disasm(code, pc))
 
-        #PC points to symbolic memory
-        if instruction.size > len(text):
-            logger.info("Trying to execute instructions from invalid memory")
-            raise InvalidPCException(pc)
+        #decode the instructtion from code 
+        try:
+            instruction = next(self._md.disasm(code, pc))
+        except StopIteration as e:
+            raise DecodeException(pc, code) 
+
 
+        #Check that the decoded intruction is contained in executable memory
         if not self.memory.access_ok(slice(pc, pc+instruction.size), 'x'):
             logger.info("Trying to execute instructions from non-executable memory")
-            raise InvalidPCException(pc)
+            raise InvalidMemoryAccess(pc, 'x')
 
         instruction.operands = self._wrap_operands(instruction.operands)
 
         self._instruction_cache[pc] = instruction
-        self.instruction = instruction
+        return instruction
 
+    @property
+    def instruction(self):
+        if self._last_pc is None:
+            return self.decode_instruction(self.PC)
+        else:
+            return self.decode_instruction(self._last_pc)
 
     #######################################
     # Execute
-    @abstractmethod
     def canonicalize_instruction_name(self, instruction):
         '''
         Get the semantic name of an instruction.
         '''
-        pass
+        raise NotImplemented
 
     def execute(self):
         '''
         Decode, and execute one instruction pointed by register PC
         '''
+        if issymbolic(self.PC):
+            raise ConcretizeRegister(self, 'PC', policy='ALL')
+
+        if not self.memory.access_ok(self.PC,'x'):
+            raise InvalidMemoryAccess(self.PC, 'x')
+
+        self.will_decode_instruction()
```

Comment:
Problem is that cpu.instruction is a property that decodes from where the PC points to at that moment. And it may be used in odd places. All signal emits are in question. Lets do a pass after merge and maybe after used a bit.